src/bin/basic_subclass.rs

```diff
@@ -0,0 +1,185 @@
+}
+
+impl ObjectSubclass for SimpleWindowPrivate {
+    const NAME: &'static str = "SimpleWindowPrivate";
+    type ParentType = gtk::ApplicationWindow;
+    type Instance = subclass::simple::InstanceStruct<Self>;
+    type Class = subclass::simple::ClassStruct<Self>;
+
+    glib_object_subclass!();
+
+    fn new() -> Self {
+        Self {
+            widgets: OnceCell::new(),
+        }
+    }
+}
+
+impl ObjectImpl for SimpleWindowPrivate {
+    glib_object_impl!();
+
+    fn constructed(&self, obj: &glib::Object) {
+        self.parent_constructed(obj);
+        let self_ = obj.downcast_ref::<SimpleWindow>().unwrap();
+
+        let headerbar = gtk::HeaderBar::new();
+        headerbar.set_title(Some("Hello World!"));
+        headerbar.set_show_close_button(true);
+
+        let label = gtk::Label::new(Some("Hello World!"));
+
+        self_.add(&label);
+        self_.set_titlebar(Some(&headerbar));
+        self_.set_default_size(640, 480);
+
+        self.widgets
+            .set(WindowWidgets { headerbar, label })
+            .expect("Failed to initialize window state");
+    }
+}
+
+impl WidgetImpl for SimpleWindowPrivate {}
+impl ContainerImpl for SimpleWindowPrivate {}
+impl BinImpl for SimpleWindowPrivate {}
+impl WindowImpl for SimpleWindowPrivate {}
+impl ApplicationWindowImpl for SimpleWindowPrivate {}
+
+glib_wrapper! {
+    pub struct SimpleWindow(
+        Object<subclass::simple::InstanceStruct<SimpleWindowPrivate>,
+        subclass::simple::ClassStruct<SimpleWindowPrivate>,
+        SimpleAppWindowClass>)
+        @extends gtk::Widget, gtk::Container, gtk::Bin, gtk::Window, gtk::ApplicationWindow;
+
+    match fn {
+        get_type => || SimpleWindowPrivate::get_type().to_glib(),
+    }
+}
+
+impl SimpleWindow {
+    pub fn new(app: &gtk::Application) -> Self {
+        glib::Object::new(Self::static_type(), &[("application", app)])
+            .expect("Failed to create SimpleWindow")
+            .downcast::<SimpleWindow>()
+            .expect("Created SimpleWindow is of wrong type")
+    }
+}
+
+#[derive(Debug)]
+struct State {
+    window: SimpleWindow,
+}
+
+pub struct SimpleApplicationPrivate {
+    state: OnceCell<State>,
+}
+
+impl ObjectSubclass for SimpleApplicationPrivate {
+    const NAME: &'static str = "SimpleApplicationPrivate";
+    type ParentType = gtk::Application;
+    type Instance = subclass::simple::InstanceStruct<Self>;
+    type Class = subclass::simple::ClassStruct<Self>;
+
+    glib_object_subclass!();
+
+    fn new() -> Self {
+        Self {
+            state: OnceCell::new(),
+        }
+    }
+}
+
+impl ObjectImpl for SimpleApplicationPrivate {
+    glib_object_impl!();
+}
+
+impl ApplicationImpl for SimpleApplicationPrivate {
+    fn activate(&self, app: &gio::Application) {
+        let app = app.downcast_ref::<gtk::Application>().unwrap();
+        let priv_ = SimpleApplicationPrivate::from_instance(app);
+        if let Some(ref s) = priv_.state.get() {
```

Comment:
```suggestion
        if let Some(ref s) = priv_.state.get().expect("activate called before constructed") {
```

Reply:
Or not?

Reply:
If you want to keep the current behaviour, you maybe want to use `priv_.state.get_or_init(|| initalize_things)`

Reply:
I think you generally want different behavior on initializing vs activating so I don't think `.get_or_init` would work as you'd not make use of your state afterwards.

Reply:
Maybe? :)

```rust
let state = priv_.state.get_or_init(|| init_your_things);
state.window.present();
```

You might want to do more things with the state later.

Reply:
I just remembered there's a startup method we could implement instead. What do you thing about 550f437 ?

Reply:
Makes sense, yes. `show_all()` should also be moved to `activate()` though from my understanding.